File: keep-alive.py
#!/usr/bin/env python3
"""
BloomIQ Backend Keep-Alive Service
Pings the backend every 10 minutes to prevent Render free tier from sleeping
"""

import requests
import time
from datetime import datetime
import sys

# Configuration
BACKEND_URL = "https://bloomiq.onrender.com"
PING_INTERVAL = 600  # 10 minutes in seconds

# ANSI color codes
class Colors:
    RESET = '\033[0m'
    GREEN = '\033[92m'
    RED = '\033[91m'
    YELLOW = '\033[93m'
    BLUE = '\033[94m'

def ping_backend():
    """Ping the backend health and root endpoints"""
    timestamp = datetime.now().isoformat()
    print(f"\n{Colors.BLUE}[{timestamp}] 🔄 Pinging backend...{Colors.RESET}")
    
    try:
        # Ping health endpoint
        health_response = requests.get(
            f"{BACKEND_URL}/health",
            timeout=30
        )
        
        if health_response.status_code == 200:
            data = health_response.json()
            print(f"{Colors.GREEN}✅ Health check passed!{Colors.RESET}")
            print(f"{Colors.GREEN}   Status: {data.get('status')}{Colors.RESET}")
            print(f"{Colors.GREEN}   MongoDB: {data.get('mongodb')}{Colors.RESET}")
            print(f"{Colors.GREEN}   Uptime: {int(data.get('uptime', 0))} seconds{Colors.RESET}")
        else:
            print(f"{Colors.YELLOW}⚠️  Health check returned: {health_response.status_code}{Colors.RESET}")
        
        # Ping root endpoint
        root_response = requests.get(f"{BACKEND_URL}/", timeout=30)
        if root_response.status_code == 200:
            print(f"{Colors.GREEN}✅ Root endpoint check passed!{Colors.RESET}")
        
        return True
        
    except requests.exceptions.Timeout:
        print(f"{Colors.YELLOW}⚠️  Request timeout (backend might be waking up){Colors.RESET}")
        return False
    except requests.exceptions.RequestException as e:
        print(f"{Colors.RED}❌ Error pinging backend: {str(e)}{Colors.RESET}")
        return False

def main():
    """Main function to run the keep-alive service"""
    print(f"{Colors.BLUE}🚀 BloomIQ Backend Keep-Alive Service{Colors.RESET}")
    print(f"{Colors.BLUE}{'=' * 50}{Colors.RESET}")
    print(f"{Colors.BLUE}Backend URL: {BACKEND_URL}{Colors.RESET}")
    print(f"{Colors.BLUE}Interval: Every {PING_INTERVAL // 60} minutes{Colors.RESET}")
    print(f"{Colors.BLUE}Started at: {datetime.now().isoformat()}{Colors.RESET}")
    print(f"{Colors.BLUE}{'=' * 50}{Colors.RESET}\n")
    
    # Run immediately on start
    ping_backend()
    
    # Run in loop
    try:
        while True:
            time.sleep(PING_INTERVAL)
            ping_backend()
    except KeyboardInterrupt:
        print(f"\n{Colors.YELLOW}👋 Shutting down keep-alive service...{Colors.RESET}")
        sys.exit(0)

if __name__ == "__main__":
    main()

File: keep-alive.js
const cron = require('node-cron');
const axios = require('axios');

// Configuration
const BACKEND_URL = 'https://bloomiq.onrender.com';
const PING_INTERVAL = '*/10 * * * *'; // Every 10 minutes

// Color codes for console
const colors = {
  reset: '\x1b[0m',
  green: '\x1b[32m',
  red: '\x1b[31m',
  yellow: '\x1b[33m',
  blue: '\x1b[34m'
};

// Ping function
async function pingBackend() {
  const timestamp = new Date().toISOString();
  console.log(`\n${colors.blue}[${timestamp}] 🔄 Pinging backend...${colors.reset}`);

  try {
    // Ping health endpoint
    const healthResponse = await axios.get(`${BACKEND_URL}/health`, {
      timeout: 30000 // 30 seconds timeout
    });

    if (healthResponse.status === 200) {
      console.log(`${colors.green}✅ Health check passed!${colors.reset}`);
      console.log(`${colors.green}   Status: ${healthResponse.data.status}${colors.reset}`);
      console.log(`${colors.green}   MongoDB: ${healthResponse.data.mongodb}${colors.reset}`);
      console.log(`${colors.green}   Uptime: ${Math.floor(healthResponse.data.uptime)} seconds${colors.reset}`);
    }

    // Ping root endpoint
    const rootResponse = await axios.get(`${BACKEND_URL}/`);
    console.log(`${colors.green}✅ Root endpoint check passed!${colors.reset}`);

    return true;
  } catch (error) {
    if (error.code === 'ECONNABORTED') {
      console.log(`${colors.yellow}⚠️  Request timeout (backend might be waking up)${colors.reset}`);
    } else if (error.response) {
      console.log(`${colors.red}❌ Backend returned error: ${error.response.status}${colors.reset}`);
    } else if (error.request) {
      console.log(`${colors.red}❌ No response from backend${colors.reset}`);
    } else {
      console.log(`${colors.red}❌ Error: ${error.message}${colors.reset}`);
    }
    return false;
  }
}

// Start cron job
console.log(`${colors.blue}🚀 BloomIQ Backend Keep-Alive Service${colors.reset}`);
console.log(`${colors.blue}${'='.repeat(50)}${colors.reset}`);
console.log(`${colors.blue}Backend URL: ${BACKEND_URL}${colors.reset}`);
console.log(`${colors.blue}Interval: Every 10 minutes${colors.reset}`);
console.log(`${colors.blue}Started at: ${new Date().toISOString()}${colors.reset}`);
console.log(`${colors.blue}${'='.repeat(50)}${colors.reset}\n`);

// Run immediately on start
pingBackend();

// Schedule recurring pings
cron.schedule(PING_INTERVAL, () => {
  pingBackend();
});

// Keep process running
process.on('SIGINT', () => {
  console.log(`\n${colors.yellow}👋 Shutting down keep-alive service...${colors.reset}`);
  process.exit(0);
});

// Handle uncaught errors
process.on('uncaughtException', (error) => {
  console.error(`${colors.red}❌ Uncaught Exception: ${error.message}${colors.reset}`);
});

process.on('unhandledRejection', (reason, promise) => {
  console.error(`${colors.red}❌ Unhandled Rejection at:${colors.reset}`, promise);
  console.error(`${colors.red}   Reason:${colors.reset}`, reason);
});
